### ISSUE TITLE
counter_cache doesn't change when change parent if parent and child have same class

### ISSUE THREAD

**Issue by senid231** [rails/rails#31186]
### Steps to reproduce
```ruby
# frozen_string_literal: true

begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  # gem 'rails', github: 'rails/rails', require: false
  gem 'activerecord', '~> 5.0'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :employees, force: true do |t|
    t.string :name, null: false
    t.integer :boss_id
    t.integer :subordinates_count, null: false, default: 0
  end
end

class Employee < ActiveRecord::Base
  belongs_to :boss, class_name: 'Employee', foreign_key: :boss_id, counter_cache: :subordinates_count
  has_many :subordinates, class_name: 'Employee', foreign_key: :boss_id
end

class BugTest < Minitest::Test

  def assert_subordinates(employee_id, expected_count)
    assert_equal expected_count, Employee.find(employee_id).subordinates_count
    assert_equal expected_count, Employee.where(boss_id: employee_id).count(:all)
  end

  def test_cache_counter
    boss = Employee.create! name: 'John'
    assert_subordinates boss.id, 0

    employee = Employee.create! name: 'Dave', boss_id: boss.id
    assert_subordinates boss.id, 1

    boss_2 = Employee.create! name: 'Jane'
    assert_subordinates boss_2.id, 0

    employee.update! boss_id: boss_2.id
    assert_subordinates boss_2.id, 1
    assert_subordinates boss.id, 0
  end

  def test_cache_counter_with_clean_object
    boss = Employee.create! name: 'John'
    assert_subordinates boss.id, 0

    employee = Employee.create! name: 'Dave', boss_id: boss.id
    assert_subordinates boss.id, 1

    boss_2 = Employee.create! name: 'Jane'
    assert_subordinates boss_2.id, 0

    Employee.find(employee.id).update! boss_id: boss_2.id
    assert_subordinates boss_2.id, 1
    assert_subordinates boss.id, 0
  end
end
```
### Expected behavior
after updating changing `boss_id` from `boss.id` to `boss_2.id`
`boss` should have `subordinates_count` equal to `0`
and `boss_2` should have `subordinates_count` equal to `1`

### Actual behavior
`subordinates_count` doesn't change for `boss` and `boss_2`

### System configuration
**Rails version**: 5.1.4 
same behaviour on 5.0.6
can't start tests on `master` because `bundle install` failed

**Ruby version**: 2.3.1

**Comment by senid231** [comment]
Log
```
Testing started at 4:31 PM ...
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Using concurrent-ruby 1.0.5
Using i18n 0.9.1 (was 0.8.6)
Using minitest 5.10.3
Using thread_safe 0.3.6
Using tzinfo 1.2.4 (was 1.2.3)
Using activesupport 5.1.4 (was 5.0.6)
Using activemodel 5.1.4 (was 5.0.6)
Using arel 8.0.0 (was 7.1.4)
Using activerecord 5.1.4 (was 5.0.6)
Using bundler 1.16.0
Using sqlite3 1.3.13
-- create_table(:employees, {:force=>true})
D, [2017-11-20T16:31:08.160586 #4607] DEBUG -- :    (0.1ms)  DROP TABLE IF EXISTS "employees"
D, [2017-11-20T16:31:08.161948 #4607] DEBUG -- :    (0.6ms)  SELECT sqlite_version(*)
D, [2017-11-20T16:31:08.162285 #4607] DEBUG -- :    (0.2ms)  CREATE TABLE "employees" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "name" varchar NOT NULL, "boss_id" integer, "subordinates_count" integer DEFAULT 0 NOT NULL)
   -> 0.0044s
D, [2017-11-20T16:31:08.178329 #4607] DEBUG -- :    (0.2ms)  CREATE TABLE "ar_internal_metadata" ("key" varchar NOT NULL PRIMARY KEY, "value" varchar, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL)
D, [2017-11-20T16:31:08.185881 #4607] DEBUG -- :   ActiveRecord::InternalMetadata Load (0.2ms)  SELECT  "ar_internal_metadata".* FROM "ar_internal_metadata" WHERE "ar_internal_metadata"."key" = ? LIMIT ?  [["key", "environment"], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.189371 #4607] DEBUG -- :    (0.0ms)  begin transaction
D, [2017-11-20T16:31:08.190352 #4607] DEBUG -- :   SQL (0.1ms)  INSERT INTO "ar_internal_metadata" ("key", "value", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["key", "environment"], ["value", "test"], ["created_at", "2017-11-20 14:31:08.189685"], ["updated_at", "2017-11-20 14:31:08.189685"]]
D, [2017-11-20T16:31:08.190568 #4607] DEBUG -- :    (0.0ms)  commit transaction
Started

D, [2017-11-20T16:31:08.203276 #4607] DEBUG -- :    (0.1ms)  begin transaction
D, [2017-11-20T16:31:08.204173 #4607] DEBUG -- :   SQL (0.1ms)  INSERT INTO "employees" ("name") VALUES (?)  [["name", "John"]]
D, [2017-11-20T16:31:08.206493 #4607] DEBUG -- :    (0.1ms)  commit transaction
D, [2017-11-20T16:31:08.207715 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.208642 #4607] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 1]]
D, [2017-11-20T16:31:08.208901 #4607] DEBUG -- :    (0.0ms)  begin transaction
D, [2017-11-20T16:31:08.209433 #4607] DEBUG -- :   SQL (0.1ms)  INSERT INTO "employees" ("name", "boss_id") VALUES (?, ?)  [["name", "Dave"], ["boss_id", 1]]
D, [2017-11-20T16:31:08.212436 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.213073 #4607] DEBUG -- :   SQL (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) + 1 WHERE "employees"."id" = ?  [["id", 1]]
D, [2017-11-20T16:31:08.213306 #4607] DEBUG -- :    (0.0ms)  commit transaction
D, [2017-11-20T16:31:08.213550 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.214071 #4607] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 1]]
D, [2017-11-20T16:31:08.214337 #4607] DEBUG -- :    (0.0ms)  begin transaction
D, [2017-11-20T16:31:08.214864 #4607] DEBUG -- :   SQL (0.1ms)  INSERT INTO "employees" ("name") VALUES (?)  [["name", "Jane"]]
D, [2017-11-20T16:31:08.215114 #4607] DEBUG -- :    (0.0ms)  commit transaction
D, [2017-11-20T16:31:08.215316 #4607] DEBUG -- :   Employee Load (0.0ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 3], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.215785 #4607] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 3]]
D, [2017-11-20T16:31:08.216031 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 2], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.216191 #4607] DEBUG -- :    (0.0ms)  begin transaction
D, [2017-11-20T16:31:08.217341 #4607] DEBUG -- :   SQL (0.1ms)  UPDATE "employees" SET "boss_id" = ? WHERE "employees"."id" = ?  [["boss_id", 3], ["id", 2]]
D, [2017-11-20T16:31:08.217821 #4607] DEBUG -- :   SQL (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) + 1 WHERE "employees"."id" = ?  [["id", 3]]
D, [2017-11-20T16:31:08.218122 #4607] DEBUG -- :   SQL (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) - 1 WHERE "employees"."id" = ?  [["id", 1]]
D, [2017-11-20T16:31:08.218264 #4607] DEBUG -- :    (0.0ms)  commit transaction
D, [2017-11-20T16:31:08.218513 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 3], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.218981 #4607] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 3]]
D, [2017-11-20T16:31:08.219251 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.219702 #4607] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 1]]
D, [2017-11-20T16:31:08.220099 #4607] DEBUG -- :    (0.0ms)  begin transaction
D, [2017-11-20T16:31:08.220583 #4607] DEBUG -- :   SQL (0.1ms)  INSERT INTO "employees" ("name") VALUES (?)  [["name", "John"]]
D, [2017-11-20T16:31:08.220828 #4607] DEBUG -- :    (0.0ms)  commit transaction
D, [2017-11-20T16:31:08.221047 #4607] DEBUG -- :   Employee Load (0.0ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 4], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.221650 #4607] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 4]]
D, [2017-11-20T16:31:08.222013 #4607] DEBUG -- :    (0.1ms)  begin transaction
D, [2017-11-20T16:31:08.222747 #4607] DEBUG -- :   SQL (0.1ms)  INSERT INTO "employees" ("name", "boss_id") VALUES (?, ?)  [["name", "Dave"], ["boss_id", 4]]
D, [2017-11-20T16:31:08.223523 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 4], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.224495 #4607] DEBUG -- :   SQL (0.3ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) + 1 WHERE "employees"."id" = ?  [["id", 4]]
D, [2017-11-20T16:31:08.224808 #4607] DEBUG -- :    (0.0ms)  commit transaction
D, [2017-11-20T16:31:08.225088 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 4], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.225688 #4607] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 4]]
D, [2017-11-20T16:31:08.225949 #4607] DEBUG -- :    (0.0ms)  begin transaction
D, [2017-11-20T16:31:08.226578 #4607] DEBUG -- :   SQL (0.1ms)  INSERT INTO "employees" ("name") VALUES (?)  [["name", "Jane"]]
D, [2017-11-20T16:31:08.226892 #4607] DEBUG -- :    (0.0ms)  commit transaction
D, [2017-11-20T16:31:08.227170 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]
D, [2017-11-20T16:31:08.227809 #4607] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 6]]
D, [2017-11-20T16:31:08.227991 #4607] DEBUG -- :    (0.0ms)  begin transaction
D, [2017-11-20T16:31:08.228801 #4607] DEBUG -- :   SQL (0.1ms)  UPDATE "employees" SET "boss_id" = ? WHERE "employees"."id" = ?  [["boss_id", 6], ["id", 5]]
D, [2017-11-20T16:31:08.229041 #4607] DEBUG -- :    (0.0ms)  commit transaction
D, [2017-11-20T16:31:08.229284 #4607] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]

Minitest::Assertion: Expected: 1
  Actual: 0
/home/senid/projects/forked/rails/bug_report_test.rb:48:in `assert_subordinates'
/home/senid/projects/forked/rails/bug_report_test.rb:63:in `test_cache_counter'
Finished in 0.03124s
2 tests, 17 assertions, 1 failures, 0 errors, 0 skips

Process finished with exit code 0
```

**Comment by htanata** [comment]
Seems like this is fixed as of Active Record 5.2.0.beta2.

Here's my test script:

```
# frozen_string_literal: true

begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  # gem 'rails', github: 'rails/rails', require: false
  gem 'activerecord', '5.2.0.beta2'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :employees, force: true do |t|
    t.string :name, null: false
    t.integer :boss_id
    t.integer :subordinates_count, null: false, default: 0
  end
end

class Employee < ActiveRecord::Base
  belongs_to :boss, class_name: 'Employee', foreign_key: :boss_id, counter_cache: :subordinates_count
  has_many :subordinates, class_name: 'Employee', foreign_key: :boss_id, inverse_of: :boss
end

class BugTest < Minitest::Test

  def assert_subordinates(employee_id, expected_count)
    assert_equal expected_count, Employee.find(employee_id).subordinates_count
    assert_equal expected_count, Employee.where(boss_id: employee_id).count(:all)
  end

  def test_cache_counter
    boss = Employee.create! name: 'John'
    assert_subordinates boss.id, 0

    employee = Employee.create! name: 'Dave', boss_id: boss.id
    assert_subordinates boss.id, 1

    boss_2 = Employee.create! name: 'Jane'
    assert_subordinates boss_2.id, 0

    employee.update! boss_id: boss_2.id
    assert_subordinates boss_2.id, 1
    assert_subordinates boss.id, 0
  end

  def test_cache_counter_with_clean_object
    boss = Employee.create! name: 'John'
    assert_subordinates boss.id, 0

    employee = Employee.create! name: 'Dave', boss_id: boss.id
    assert_subordinates boss.id, 1

    boss_2 = Employee.create! name: 'Jane'
    assert_subordinates boss_2.id, 0

    Employee.find(employee.id).update! boss_id: boss_2.id
    assert_subordinates boss_2.id, 1
    assert_subordinates boss.id, 0
  end
end
```

Result:

```
$ ruby test.rb
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Using concurrent-ruby 1.0.5
Using minitest 5.11.1
Using thread_safe 0.3.6
Using arel 9.0.0
Using bundler 1.16.0
Using sqlite3 1.3.13
Using i18n 0.9.1
Using tzinfo 1.2.4
Fetching activesupport 5.2.0.beta2
Installing activesupport 5.2.0.beta2
Fetching activemodel 5.2.0.beta2
Installing activemodel 5.2.0.beta2
Fetching activerecord 5.2.0.beta2
Installing activerecord 5.2.0.beta2
-- create_table(:employees, {:force=>true})
D, [2018-01-19T18:41:26.752797 #37026] DEBUG -- :    (0.1ms)  DROP TABLE IF EXISTS "employees"
D, [2018-01-19T18:41:26.753915 #37026] DEBUG -- :    (0.8ms)  SELECT sqlite_version(*)
D, [2018-01-19T18:41:26.754309 #37026] DEBUG -- :    (0.3ms)  CREATE TABLE "employees" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "name" varchar NOT NULL, "boss_id" integer, "subordinates_count" integer DEFAULT 0 NOT NULL)
   -> 0.0040s
D, [2018-01-19T18:41:26.781325 #37026] DEBUG -- :    (0.6ms)  CREATE TABLE "ar_internal_metadata" ("key" varchar NOT NULL PRIMARY KEY, "value" varchar, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL)
D, [2018-01-19T18:41:26.788860 #37026] DEBUG -- :   ActiveRecord::InternalMetadata Load (0.1ms)  SELECT  "ar_internal_metadata".* FROM "ar_internal_metadata" WHERE "ar_internal_metadata"."key" = ? LIMIT ?  [["key", "environment"], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.793200 #37026] DEBUG -- :    (0.0ms)  begin transaction
D, [2018-01-19T18:41:26.794539 #37026] DEBUG -- :   ActiveRecord::InternalMetadata Create (0.1ms)  INSERT INTO "ar_internal_metadata" ("key", "value", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["key", "environment"], ["value", "default_env"], ["created_at", "2018-01-20 02:41:26.793548"], ["updated_at", "2018-01-20 02:41:26.793548"]]
D, [2018-01-19T18:41:26.795052 #37026] DEBUG -- :    (0.1ms)  commit transaction
Run options: --seed 18521

# Running:

D, [2018-01-19T18:41:26.812528 #37026] DEBUG -- :    (0.1ms)  begin transaction
D, [2018-01-19T18:41:26.813369 #37026] DEBUG -- :   Employee Create (0.1ms)  INSERT INTO "employees" ("name") VALUES (?)  [["name", "John"]]
D, [2018-01-19T18:41:26.816230 #37026] DEBUG -- :    (0.1ms)  commit transaction
D, [2018-01-19T18:41:26.817036 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.818177 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 1]]
D, [2018-01-19T18:41:26.818571 #37026] DEBUG -- :    (0.0ms)  begin transaction
D, [2018-01-19T18:41:26.819141 #37026] DEBUG -- :   Employee Create (0.1ms)  INSERT INTO "employees" ("name", "boss_id") VALUES (?, ?)  [["name", "Dave"], ["boss_id", 1]]
D, [2018-01-19T18:41:26.821806 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.822387 #37026] DEBUG -- :   Employee Update All (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) + 1 WHERE "employees"."id" = ?  [["id", 1]]
D, [2018-01-19T18:41:26.822724 #37026] DEBUG -- :    (0.0ms)  commit transaction
D, [2018-01-19T18:41:26.822971 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.823412 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 1]]
D, [2018-01-19T18:41:26.823646 #37026] DEBUG -- :    (0.0ms)  begin transaction
D, [2018-01-19T18:41:26.824077 #37026] DEBUG -- :   Employee Create (0.1ms)  INSERT INTO "employees" ("name") VALUES (?)  [["name", "Jane"]]
D, [2018-01-19T18:41:26.824345 #37026] DEBUG -- :    (0.0ms)  commit transaction
D, [2018-01-19T18:41:26.827304 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 3], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.827761 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 3]]
D, [2018-01-19T18:41:26.827939 #37026] DEBUG -- :    (0.0ms)  begin transaction
D, [2018-01-19T18:41:26.828557 #37026] DEBUG -- :   Employee Update (0.1ms)  UPDATE "employees" SET "boss_id" = ? WHERE "employees"."id" = ?  [["boss_id", 3], ["id", 2]]
D, [2018-01-19T18:41:26.829056 #37026] DEBUG -- :   Employee Update All (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) + 1 WHERE "employees"."id" = ?  [["id", 3]]
D, [2018-01-19T18:41:26.829393 #37026] DEBUG -- :   Employee Update All (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) - 1 WHERE "employees"."id" = ?  [["id", 1]]
D, [2018-01-19T18:41:26.829529 #37026] DEBUG -- :    (0.0ms)  commit transaction
D, [2018-01-19T18:41:26.829750 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 3], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.830146 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 3]]
D, [2018-01-19T18:41:26.830382 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.830766 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 1]]
.D, [2018-01-19T18:41:26.831032 #37026] DEBUG -- :    (0.0ms)  begin transaction
D, [2018-01-19T18:41:26.831403 #37026] DEBUG -- :   Employee Create (0.1ms)  INSERT INTO "employees" ("name") VALUES (?)  [["name", "John"]]
D, [2018-01-19T18:41:26.831697 #37026] DEBUG -- :    (0.0ms)  commit transaction
D, [2018-01-19T18:41:26.831885 #37026] DEBUG -- :   Employee Load (0.0ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 4], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.832282 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 4]]
D, [2018-01-19T18:41:26.832492 #37026] DEBUG -- :    (0.0ms)  begin transaction
D, [2018-01-19T18:41:26.832874 #37026] DEBUG -- :   Employee Create (0.1ms)  INSERT INTO "employees" ("name", "boss_id") VALUES (?, ?)  [["name", "Dave"], ["boss_id", 4]]
D, [2018-01-19T18:41:26.833264 #37026] DEBUG -- :   Employee Load (0.0ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 4], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.833667 #37026] DEBUG -- :   Employee Update All (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) + 1 WHERE "employees"."id" = ?  [["id", 4]]
D, [2018-01-19T18:41:26.833914 #37026] DEBUG -- :    (0.0ms)  commit transaction
D, [2018-01-19T18:41:26.834138 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 4], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.834530 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 4]]
D, [2018-01-19T18:41:26.834818 #37026] DEBUG -- :    (0.1ms)  begin transaction
D, [2018-01-19T18:41:26.835260 #37026] DEBUG -- :   Employee Create (0.1ms)  INSERT INTO "employees" ("name") VALUES (?)  [["name", "Jane"]]
D, [2018-01-19T18:41:26.835561 #37026] DEBUG -- :    (0.1ms)  commit transaction
D, [2018-01-19T18:41:26.835764 #37026] DEBUG -- :   Employee Load (0.0ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.836192 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 6]]
D, [2018-01-19T18:41:26.836425 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 5], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.836611 #37026] DEBUG -- :    (0.0ms)  begin transaction
D, [2018-01-19T18:41:26.837101 #37026] DEBUG -- :   Employee Update (0.1ms)  UPDATE "employees" SET "boss_id" = ? WHERE "employees"."id" = ?  [["boss_id", 6], ["id", 5]]
D, [2018-01-19T18:41:26.837560 #37026] DEBUG -- :   Employee Update All (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) + 1 WHERE "employees"."id" = ?  [["id", 6]]
D, [2018-01-19T18:41:26.837890 #37026] DEBUG -- :   Employee Update All (0.1ms)  UPDATE "employees" SET "subordinates_count" = COALESCE("subordinates_count", 0) - 1 WHERE "employees"."id" = ?  [["id", 4]]
D, [2018-01-19T18:41:26.838047 #37026] DEBUG -- :    (0.1ms)  commit transaction
D, [2018-01-19T18:41:26.838264 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.838662 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 6]]
D, [2018-01-19T18:41:26.838903 #37026] DEBUG -- :   Employee Load (0.1ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = ? LIMIT ?  [["id", 4], ["LIMIT", 1]]
D, [2018-01-19T18:41:26.839292 #37026] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "employees" WHERE "employees"."boss_id" = ?  [["boss_id", 4]]
.

Finished in 0.033377s, 59.9218 runs/s, 599.2176 assertions/s.

2 runs, 20 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by kamipo** [comment]
Thanks, I confirmed that this is fixed on master too.
